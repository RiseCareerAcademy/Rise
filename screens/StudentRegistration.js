import React from "react";
import {
  Text,
  Container,
  Content,
  Form,
  Item,
  Label,
  Input,
  Header,
  Right,
  Icon,
  Button,
  Body,
  Left,
  Title
} from "native-base";
import {
  StyleSheet,
  TouchableOpacity,
  View,
  Image,
  ActivityIndicator
} from "react-native";
import { connect } from "react-redux";
import { ImagePicker, Permissions } from "expo";

import { registerMentee } from "../actions/user.actions";
import { DOMAIN } from "../config/url";

const uuidv1 = require("uuid/v1");

export class StudentRegistration extends React.Component {
  static navigationOptions = {
    header: null,
  }

  state = {
    email: "",
    password: "",
    confirmedPassword: "",
    errors: [],
    skills: "",
    profession: "",
    name: "",
    city: "",
    state: "",
    image: null,
    lastName: "",
    zipcode: ""
  };

  handleImagePickerPress = async () => {
    const { status: cameraPerm } = await Permissions.askAsync(
      Permissions.CAMERA
    );
    const { status: cameraRollPerm } = await Permissions.askAsync(
      Permissions.CAMERA_ROLL
    );

    if (cameraPerm === "granted" && cameraRollPerm === "granted") {
      const result = await ImagePicker.launchImageLibraryAsync({
        allowsEditing: true, //Android editing only
        aspect: [4, 3] //Aspect ratio to maintain if user allowed to edit image
        // base64: true,
      });
      // this.base64 = result.base64;

      if (!result.cancelled) {
        this.setState({ image: result.uri });
      }
    }
  };

  handleEmail = text => {
    this.setState({ email: text });
  };
  handlePassword = text => {
    this.setState({ password: text });
  };
  handleConfirmedPassword = text => {
    this.setState({ confirmedPassword: text });
  };
  handleSkills = text => {
    this.setState({ skills: text });
  };
  handleProfession = text => {
    this.setState({ profession: text });
  };
  handleName = text => {
    this.setState({ name: text });
  };
  handleLastName = text => {
    this.setState({ lastName: text });
  };
  handleZipCode = text => {
    this.setState({ zipcode: text });
  };

  handleBiography = text => {
    this.setState({ biography: text });
  }

  validate = (
    email,
    password,
    confirmedPassword,
    skills,
    profession,
    name,
    lastName,
    zipcode,
    biography,
  ) => {
    // we are going to store errors for all fields
    // in a signle array
    const errors = [];
    if (
      skills.length == 0 ||
      profession.length == 0 ||
      name.length == 0 ||
      email.length == 0 ||
      password.length == 0 ||
      confirmedPassword.length == 0 ||
      lastName.length == 0 ||
      zipcode.length == 0 ||
      biography.length == 0
    ) {
      errors.push("All fields must be filled");
    } else if (email.length < 5) {
      errors.push("Email should be at least 5 charcters long");
    } else if (email.split("").filter(x => x === "@").length !== 1) {
      errors.push("Email should contain one @");
    } else if (email.indexOf(".") === -1) {
      errors.push("Email should contain at least one dot");
    } else if (password.length < 6) {
      errors.push("Password should be at least 6 characters long");
    } else if (password != confirmedPassword) {
      errors.push(
        "Password doesn't match" + password + " " + confirmedPassword
      );
    } else if (zipcode.length != 5 && /^\d+$/.test(zipcode)) {
      errors.push("zipcode must contain only numbers and be 5 characters long");
    }
    if (errors.length == 0) {
      // alert(errors);
      return true;
    } else {
      return false;
    }
  };

  handleSubmit = () => {
    const valid = this.validate(
      this.state.email,
      this.state.password,
      this.state.confirmedPassword,
      this.state.skills,
      this.state.profession,
      this.state.name,
      this.state.lastName,
      this.state.zipcode,
      this.state.biography,
    );
    if (!valid && process.env.NODE_ENV !== "development") {
      return;
    }

    const mentee = {
      first_name: this.state.name,
      last_name: this.state.lastName,
      email_address: this.state.email,
      biography: "hi",
      zipcode: this.state.zipcode,
      date_of_birth: "12/24/1996",
      skills: this.state.skills,
      hobbies: "fake hobbies",
      profession: this.state.profession,
      password: this.state.password,
      city: this.state.city,
      state: this.state.state,
      image: this.state.image,
      uri: this.state.image
    };

    this.props.registerMentee(mentee);
  };

  handlePreviewPress = () => {
    this.props.navigation.navigate("Preview", {
      profile_pic_URL: this.state.image,
      first_name: this.state.name,
      last_name: this.state.lastName,
      zipcode: this.state.zipcode,
      biography: this.state.biography,
      profession: this.state.profession,
      skills: this.state.skills,
      preview: true
    });
  };

  handleBackPress = () => {
    this.props.navigation.goBack();
  }

  render() {
    return (
      <Container >
        <Header>
          <Left>
            <Button transparent onPress={this.handleBackPress}>
              <Icon name="md-arrow-round-back" />
            </Button>
          </Left>
          <Body>
            <Title>Student Registration</Title>
          </Body>
          <Right>
            <Button transparent onPress={this.handlePreviewPress}>
              <Icon name="eye" />
            </Button>
          </Right>
        </Header>
        <Content>
          <Text style={styles.error}>
            {!this.props.registering && this.props.error}
          </Text>
          <Form>
            {this.state.image !== null && (
              <Image
                style={styles.userImage}
                source={{
                  uri: this.state.image
                }}
              />
            )}

            <View style={styles.uploadBtnContainer}>
              <Button
                onPress={this.handleImagePickerPress}
                style={styles.uploadBtn}
              >
                <Text>Select image from Camera Roll</Text>
              </Button>
            </View>
            <Item stackedLabel>
              <Label>Email</Label>
              <Input
                autoCapitalize="none"
                placeholder="Enter your email"
                onChangeText={this.handleEmail}
              />
            </Item>
            <Item stackedLabel>
              <Label>Password</Label>
              <Input
                autoCapitalize="none"
                placeholder="Enter your password"
                onChangeText={this.handlePassword}
                secureTextEntry={true}
              />
            </Item>
            <Item stackedLabel>
              <Label>Confirm Password</Label>
              <Input
                autoCapitalize="none"
                placeholder="Confirm Password Change"
                onChangeText={this.handleConfirmedPassword}
                secureTextEntry={true}
              />
            </Item>
            <Item stackedLabel last>
              <Label>Name</Label>
              <Input
                placeholder="Enter your name"
                onChangeText={this.handleName}
              />
            </Item>
            <Item stackedLabel last>
              <Label>lastName</Label>
              <Input
                placeholder="Enter your last name"
                onChangeText={this.handleLastName}
              />
            </Item>
            <Item stackedLabel>
              <Label>Skills</Label>
              <Input
                placeholder="Enter skills you want to learn"
                onChangeText={this.handleSkills}
              />
            </Item>
            <Item stackedLabel last>
              <Label>Profession</Label>
              <Input
                placeholder="Enter profession you want to learn"
                onChangeText={this.handleProfession}
              />
            </Item>
            <Item stackedLabel last>
              <Label>zipcode</Label>
              <Input
                placeholder="Enter your zipcode"
                onChangeText={this.handleZipCode}
              />
            </Item>
            <Item stackedLabel>
              <Label>Biography</Label>
              <Input
                placeholder="Enter your bio"
                onChangeText={this.handleBiography}
              />
            </Item>
          </Form>
        </Content>
        {this.props.registering ? (
          <ActivityIndicator animating size="large" />
        ) : (
          <TouchableOpacity
            style={styles.submitButton}
            onPress={this.handleSubmit}
          >
            <Text style={styles.submitButtonText}> Next </Text>
          </TouchableOpacity>
        )}
      </Container>
    );
  }
}

const styles = StyleSheet.create({
  container: {
    paddingTop: 23
  },
  userImage: {
    borderRadius: 85,
    borderWidth: 3,
    height: 170,
    marginBottom: 15,
    width: 170
  },
  input: {
    margin: 15,
    height: 40,
    borderColor: "#000000",
    borderWidth: 1
  },
  submitButton: {
    backgroundColor: "#000000",
    padding: 10,
    margin: 15,
    height: 40
  },
  submitButtonText: {
    color: "white"
  },
  buttonContainer: {
    flex: 1,
    flexDirection: "row",
    alignItems: "center",
    justifyContent: "center"
  },
  center: {
    flex: 1,
    flexDirection: "row",
    justifyContent: "center"
  },
  buttonStyle: {
    margin: 10
  },
  greyText: {
    color: "grey"
  },
  contentContainer: {
    paddingTop: 30
  },
  container: {
    margin: 5
  },
  uploadBtnContainer: {
    margin: "auto"
  },
  uploadBtn: {
    margin: "auto"
  },
  error: {
    color: "red"
  }
});

const mapStateToProps = state => ({
  loggedIn: state.user.loggedIn,
  registering: state.user.registering,
  error: state.user.error
});

export default connect(
  mapStateToProps,
  { registerMentee }
)(StudentRegistration);
