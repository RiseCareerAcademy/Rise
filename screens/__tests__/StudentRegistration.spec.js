import "react-native";
import React from "react";
import { StudentRegistration } from "../StudentRegistration";
import { shallow } from "enzyme";
global.alert = jest.fn();

it("returns true if email state is changed", () => {
  const props = {
    navigation: {
      navigate: () => {},
    },
  };
  const instance = shallow(<StudentRegistration {...props} />).instance();
  instance.handleEmail("user@email.com");
  expect(instance.state.email).toEqual("user@email.com");
});

it("returns true if password state is changed", () => {
  const props = {
    navigation: {
      navigate: () => {},
    },
  };
  const instance = shallow(<StudentRegistration {...props} />).instance();
  instance.handlePassword("password");
  expect(instance.state.password).toEqual("password");
});

it("returns true if confirmPassword state is changed", () => {
  const props = {
    navigation: {
      navigate: () => {},
    },
  };
  const instance = shallow(<StudentRegistration {...props} />).instance();
  instance.handleConfirmedPassword("password");
  expect(instance.state.confirmedPassword).toEqual("password");
});

it("returns true if skill state is changed", () => {
  const props = {
    navigation: {
      navigate: () => {},
    },
  };
  const instance = shallow(<StudentRegistration {...props} />).instance();
  instance.handleSkills("skill");
  expect(instance.state.skills).toEqual("skill");
});

it("returns true if profession state is changed", () => {
  const props = {
    navigation: {
      navigate: () => {},
    },
  };
  const instance = shallow(<StudentRegistration {...props} />).instance();
  instance.handleProfession("profession");
  expect(instance.state.profession).toEqual("profession");
});

it("returns true if name state is changed", () => {
  const props = {
    navigation: {
      navigate: () => {},
    },
  };
  const instance = shallow(<StudentRegistration {...props} />).instance();
  instance.handleName("name");
  expect(instance.state.name).toEqual("name");
});

it("returns error if all fields are not filled", () => {
  const props = {
    navigation: {
      navigate: () => {},
    },
  };
  const instance = shallow(<StudentRegistration {...props} />).instance();
  expect(
    instance.validate(
      "",
      "password",
      "password",
      "skills",
      "profession",
      "name",
      "lastName",
      "44444",
      "biography",
      "image"
    )
  ).toEqual(false);
});

it("returns error if email length is less than 5 characters long", () => {
  const props = {
    navigation: {
      navigate: () => {},
    },
  };
  const instance = shallow(<StudentRegistration {...props} />).instance();
  expect(
    instance.validate(
      "u@.c",
      "password",
      "password",
      "skills",
      "profession",
      "name",
      "lastName",
      "44444",
      "biography",
      "image"
    )
  ).toEqual(false);
});

it("returns error if email doesnt have one @", () => {
  const props = {
    navigation: {
      navigate: () => {},
    },
  };
  const instance = shallow(<StudentRegistration {...props} />).instance();
  expect(
    instance.validate(
      "u@.@@c",
      "password",
      "password",
      "skills",
      "profession",
      "name",
      "lastName",
      "44444",
      "biography",
      "image"
    )
  ).toEqual(false);
});

it("returns error if email doesnt have atleast one .", () => {
  const props = {
    navigation: {
      navigate: () => {},
    },
  };
  const instance = shallow(<StudentRegistration {...props} />).instance();
  expect(
    instance.validate(
      "user@emailcom",
      "password",
      "password",
      "skills",
      "profession",
      "name",
      "lastName",
      "44444",
      "biography",
      "image"
    )
  ).toEqual(false);
});

it("returns error if password length is less than 6", () => {
  const props = {
    navigation: {
      navigate: () => {},
    },
  };
  const instance = shallow(<StudentRegistration {...props} />).instance();
  expect(
    instance.validate(
      "user@email.com",
      "pass",
      "pass",
      "skills",
      "profession",
      "name",
      "lastName",
      "44444",
      "biography",
      "image"
    )
  ).toEqual(false);
});

it("returns error if password doesn't match confirmPassword", () => {
  const props = {
    navigation: {
      navigate: () => {},
    },
  };
  const instance = shallow(<StudentRegistration {...props} />).instance();
  expect(
    instance.validate(
      "user@email.com",
      "passwoord",
      "password",
      "skills",
      "profession",
      "name",
      "lastName",
      "44444",
      "biography",
      "image"
    )
  ).toEqual(false);
});

it("renders correctly", () => {
  const props = {
    navigation: {
      navigate: () => {},
    },
  };
  const wrapper = shallow(
    <StudentRegistration {...props}>Snapshot test!</StudentRegistration>
  );
  expect(wrapper).toMatchSnapshot();
});
