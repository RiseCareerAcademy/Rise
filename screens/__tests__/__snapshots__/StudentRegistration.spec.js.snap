// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<Styled(Container)>
  <Styled(Header)>
    <Styled(Left)>
      <Styled(Button)
        onPress={[Function]}
        transparent={true}
      >
        <Styled(Icon)
          name="md-arrow-round-back"
        />
      </Styled(Button)>
    </Styled(Left)>
    <Styled(Body)>
      <Styled(Title)>
        Student Registration
      </Styled(Title)>
    </Styled(Body)>
    <Styled(Right)>
      <Styled(Button)
        onPress={[Function]}
        transparent={true}
      >
        <Styled(Icon)
          name="eye"
        />
      </Styled(Button)>
    </Styled(Right)>
  </Styled(Header)>
  <Styled(Content)>
    <Styled(Text)
      style={
        Object {
          "color": "red",
        }
      }
    />
    <Styled(Form)>
      <Component
        style={
          Object {
            "margin": "auto",
          }
        }
      >
        <Styled(Button)
          onPress={[Function]}
          style={
            Object {
              "margin": "auto",
            }
          }
        >
          <Styled(Text)>
            Select image from Camera Roll
          </Styled(Text)>
        </Styled(Button)>
      </Component>
      <Styled(Item)
        stackedLabel={true}
      >
        <Styled(Label)>
          Email
        </Styled(Label)>
        <Styled(Input)
          autoCapitalize="none"
          keyboardType="email-address"
          onChangeText={[Function]}
          placeholder="Enter your email"
        />
      </Styled(Item)>
      <Styled(Item)
        stackedLabel={true}
      >
        <Styled(Label)>
          Password
        </Styled(Label)>
        <Styled(Input)
          autoCapitalize="none"
          onChangeText={[Function]}
          placeholder="Enter your password"
          secureTextEntry={true}
        />
      </Styled(Item)>
      <Styled(Item)
        stackedLabel={true}
      >
        <Styled(Label)>
          Confirm Password
        </Styled(Label)>
        <Styled(Input)
          autoCapitalize="none"
          onChangeText={[Function]}
          placeholder="Confirm Password Change"
          secureTextEntry={true}
        />
      </Styled(Item)>
      <Styled(Item)
        last={true}
        stackedLabel={true}
      >
        <Styled(Label)>
          Name
        </Styled(Label)>
        <Styled(Input)
          onChangeText={[Function]}
          placeholder="Enter your name"
        />
      </Styled(Item)>
      <Styled(Item)
        last={true}
        stackedLabel={true}
      >
        <Styled(Label)>
          Last Name
        </Styled(Label)>
        <Styled(Input)
          onChangeText={[Function]}
          placeholder="Enter your last name"
        />
      </Styled(Item)>
      <Styled(Item)
        stackedLabel={true}
      >
        <Styled(Label)>
          Skills (separated by commas)
        </Styled(Label)>
        <Styled(Input)
          onChangeText={[Function]}
          placeholder="Enter skills you want to learn"
        />
      </Styled(Item)>
      <Styled(Item)
        last={true}
        stackedLabel={true}
      >
        <Styled(Label)>
          Profession
        </Styled(Label)>
        <Styled(Input)
          onChangeText={[Function]}
          placeholder="Enter profession you want to learn"
        />
      </Styled(Item)>
      <Styled(Item)
        last={true}
        stackedLabel={true}
      >
        <Styled(Label)>
          zipcode
        </Styled(Label)>
        <Styled(Input)
          onChangeText={[Function]}
          placeholder="Enter your zipcode"
        />
      </Styled(Item)>
      <Styled(Item)
        stackedLabel={true}
      >
        <Styled(Label)>
          Biography
        </Styled(Label)>
        <Styled(Input)
          onChangeText={[Function]}
          placeholder="Enter your bio"
        />
      </Styled(Item)>
    </Styled(Form)>
  </Styled(Content)>
  <TouchableOpacity
    activeOpacity={0.2}
    onPress={[Function]}
    style={
      Object {
        "backgroundColor": "#000000",
        "height": 40,
        "margin": 15,
        "padding": 10,
      }
    }
  >
    <Styled(Text)
      style={
        Object {
          "color": "white",
        }
      }
    >
       Next 
    </Styled(Text)>
  </TouchableOpacity>
</Styled(Container)>
`;
